java-checks/src/main/java/org/sonar/java/checks/spring/SpelExpressionCheck.java

```diff
@@ -0,0 +1,246 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.spring;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.ObjIntConsumer;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.java.model.DefaultJavaFileScannerContext;
+import org.sonar.java.reporting.AnalyzerMessage;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.location.Position;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.springframework.expression.ParseException;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+
+@Rule(key = "S6857")
+public class SpelExpressionCheck extends IssuableSubscriptionVisitor {
+
+  private static final String SPRING_PREFIX = "org.springframework";
+
+  private static final Pattern PROPERTY_PLACEHOLDER_PATTERN = Pattern.compile(
+    "[a-zA-Z_]\\w*+(\\[\\d++])*+(\\.[a-zA-Z_]\\w*+(\\[\\d++])*+)*+"
+  );
+
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CLASS, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    getClassAndMemberAnnotations((ClassTree) tree)
+      .filter(SpelExpressionCheck::isSpringAnnotation)
+      .forEach(this::checkSpringAnnotationArguments);
+  }
+
+  private static Stream<AnnotationTree> getClassAndMemberAnnotations(ClassTree cls) {
+    return Stream.concat(
+      Stream.of(cls.modifiers().annotations()),
+      cls.members().stream().map(SpelExpressionCheck::getMemberAnnotations)
+    ).flatMap(Collection::stream);
+  }
+
+  private static List<AnnotationTree> getMemberAnnotations(Tree member) {
+    if (member.is(Tree.Kind.METHOD)) {
+      return ((MethodTree) member).modifiers().annotations();
+    } else if (member.is(Tree.Kind.VARIABLE)) {
+      return ((VariableTree) member).modifiers().annotations();
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  private static boolean isSpringAnnotation(AnnotationTree annotation) {
+    return annotation.symbolType().fullyQualifiedName().startsWith(SPRING_PREFIX);
+  }
+
+  private void checkSpringAnnotationArguments(AnnotationTree annotation) {
+    annotation.arguments().stream()
+      .filter(argument -> argument.is(Tree.Kind.STRING_LITERAL))
```

Comment:
There is also another limitation here which is that a SpringExpression might be a resolvable constant rather than a literal.
```java
// Defined in the same class or elsewhere
public static final String MY_CONSTANT_EXPRESSION = "${1 + 2 + 3}"
// ...
@Value(MY_CONSTANT_EXPRESSION)
String myProperty;
```

Reply:
OMG. This has implications for reporting exact error position inside the string literal again. Fixed.